engine/src/main/java/org/pentaho/di/trans/steps/script/ScriptAddedFunctions.java

```diff
@@ -0,0 +1,2532 @@
+
+  // This is only used for reading, so no concurrency problems.
+  // todo: move in the real variables of the step.
+  // private static VariableSpace variables = Variables.getADefaultVariableSpace();
+
+  // Functions to Add
+  // date2num, num2date,
+  // fisc_date, isNull
+  //
+
+  public static String getDigitsOnly( ScriptEngine actualContext, Bindings actualObject, Object[] ArgList,
+    Object FunctionContext ) {
+    if ( ArgList.length == 1 ) {
+      return Const.getDigitsOnly( (String) ArgList[0] ); // TODO AKRETION ensure
+    } else {
+      throw new RuntimeException( "The function call getDigitsOnly requires 1 argument." );
+
+    }
+  }
+
+  public static boolean LuhnCheck( ScriptEngine actualContext, Bindings actualObject, Object[] ArgList,
+    Object FunctionContext ) {
+
+    boolean returnCode = false;
+
+    if ( ArgList.length == 1 ) {
+      if ( !isNull( ArgList ) && !isUndefined( ArgList ) ) {
+        try {
+          int sum = 0;
+          int digit = 0;
+          int addend = 0;
+          boolean timesTwo = false;
+          String argstring = (String) ArgList[0];
+
+          for ( int i = argstring.length() - 1; i >= 0; i-- ) {
+            digit = Integer.parseInt( argstring.substring( i, i + 1 ) );
+            if ( timesTwo ) {
+              addend = digit * 2;
+              if ( addend > 9 ) {
+                addend -= 9;
+              }
+            } else {
+              addend = digit;
+            }
+            sum += addend;
+            timesTwo = !timesTwo;
+          }
+
+          int modulus = sum % 10;
+          if ( modulus == 0 ) {
+            returnCode = true;
+          }
+        } catch ( Exception e ) {
+          // No Need to throw exception
+          // This means that input can not be parsed to Integer
+        }
+
+      }
+    } else {
+      throw new RuntimeException( "The function call LuhnCheck requires 1 argument." );
+
+    }
+    return returnCode;
+  }
+
+  public static int indexOf( ScriptEngine actualContext, Bindings actualObject, Object[] ArgList,
+    Object FunctionContext ) {
+
+    int returnIndex = -1;
+
+    if ( ArgList.length == 2 || ArgList.length == 3 ) {
+      if ( !isNull( ArgList ) && !isUndefined( ArgList ) ) {
+        String string = (String) ArgList[0];
+        String subString = (String) ArgList[1];
+
+        int fromIndex = 0;
+        if ( ArgList.length == 3 ) {
+          fromIndex = (int) Math.round( (Double) ArgList[2] );
+        }
+        returnIndex = string.indexOf( subString, fromIndex );
+      }
+    } else {
+      throw new RuntimeException( "The function call indexOf requires 2 or 3 arguments" );
+    }
+    return returnIndex;
+  }
+
+  public static Object getTransformationName( ScriptEngine actualContext, Bindings actualObject, Object[] ArgList,
+    Object FunctionContext ) {
+    try {
+      Object objTranName = actualObject.get( "_TransformationName_" );
+      return objTranName;
+    } catch ( Exception e ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)